### ISSUE TITLE
Strange behavior for multilib

### ISSUE THREAD

**Issue by j-mracek** [rpm-software-management/rpm#2837]
The issue was originally reported on DNF5.

The problem is that upgrade of `B-1-1.i686` result in RPM in removal of `B-2-2.x86_64` and `B-1-1.x86_64`.

The issue is reproducible from commandline, API (Python  DNF, C DNF5). The behavior caused for DNF4 complains in callbacks, but DNF5 correctly fails. Not nicely, but it is less problematic then removal of all packages `b.x86_64`. 

Package `B` does not require anything and provides only `B`.

```
#rpm -q --root=/tmp/dnf_ci_installroot_eajo9b5c B
B-1-1.i686
B-2-2.x86_64
B-1-1.x86_64

# rpm -U --root=/tmp/dnf_ci_installroot_eajo9b5c B-2-2.i686.rpm

# rpm -q --root=/tmp/dnf_ci_installroot_eajo9b5c B
B-2-2.i686
```

_Originally posted by @j-mracek in https://github.com/rpm-software-management/dnf5/issues/518#issuecomment-1883115076_

**Comment by pmatilai** [comment]
Rpm erasing the older versions is more or less expected (see below), but that the equal version too seems bizarre and buggy alright. Can you provide the reproducer for B, as simple as it may be?

For the erasure of older versions, the behavior depends on the package colors, not architecture, although the two are related. So typically if you create an empty test package that differs by arch only, you see this behavior but if there are ELF files present (so it's "colored"), only the matching color (think arch) package is removed on upgrade.

**Comment by j-mracek** [comment]
I used two packages from our CI - https://github.com/rpm-software-management/ci-dnf-stack/tree/main/dnf-behave-tests/fixtures/specs/security-upgrade-multilib

```
Name:           B
Epoch:          0
Version:        1
Release:        1

License:        Public Domain
URL:            None

Summary:        Testing advisory upgrade options

%description
This package is part of testing security options

%files

%changelog
```
```
Name:           B
Epoch:          0
Version:        2
Release:        2

License:        Public Domain
URL:            None

Summary:        Testing advisory upgrade options

%description
This package is part of testing security options

%files

%changelog
```

**Comment by pmatilai** [comment]
Easily reproduced with that. This is not just a bug but a regression too, 4.14 doesn't behave this way.

**Comment by pmatilai** [comment]
Bisecting points the regression to commit 21836bc7524f8fc07972e0e56eed1c3b68775368

**Comment by mlschroe** [comment]
AFAIR if the packages contain no files, they wont get any color.

**Comment by mlschroe** [comment]
Do we really want to do something special for "same NEVR"? I find that a bit weird.

(But all that color handling is also super weird to me, and dnf/zypper also does not know the package colors when calculating the transaction and has to hope that the architecture matches the color.)

(BTW, is the TE color really only set by rpmteColorDS() which goes over the requires/provides dependencies and not directly from the files?)

**Comment by pmatilai** [comment]
No colors are involved in this case (no files, but yes rpmteColorDS() is the key to the coloring thing. It's weird alright.

I guess I thought thought the patch only affected --replacepkgs behavior when that's what the commit message talks about and that's the testcase that chances as well. But I don't think an upgrade with no special flags should silently replace an identical NEVR.

**Comment by j-mracek** [comment]
Be honest from API and DNF side we would love to see that RPM will not modify transaction request, but only check whether it is applicable.

**Comment by mlschroe** [comment]
What's so special about an identical NEVR? The content may still be different, we had a couple of bug reports about same NEVR with different files and rpm not doing the correct thing.

**Comment by pmatilai** [comment]
> What's so special about an identical NEVR? The content may still be different, we had a couple of bug reports about same NEVR with different files and rpm not doing the correct thing.

Because by the very definition of rpm, release must change any time content changes?
That people short-circuit and "cheat" about that in local test builds is a special case that IMO warrants using a special flag/op as well.

"Right thing" wrt this seems as subjective as anything. If you have references to those bugs I'd be interested, but what happens in this scenario doesn't seem right to me either.

**Comment by mlschroe** [comment]
An example would be: https://bugzilla.redhat.com/show_bug.cgi?id=966715

Don't you think an architecture change from i585 to i686 should delete the i586 package even if the NEVR is the same?

**Comment by pmatilai** [comment]
Right, that one. An architecture change is a fairly special operation, I'm not convinced it needs to "just work" with regular update, it just needs to be doable. Eg, using --replacepkgs, which back then it wasn't. Also, an i686 package *can* be considered an update to an i586 package, but a i686 package is not an upgrade to x86_64 one by any stretch of imagination.

Of course the starting situation here is an abnormal one, but I'd rather get an error than the outcome we now get.

**Comment by arekm** [comment]
I'm playing with claude on exactly the same issue on PLD/linux rpm 6.0.1. But also checked Fedora 42 and current Fedora rawhide:

```
$ cat colortest.spec
Name:    colortest
Version: 1.0
Release: 1
Summary: Uncolored multilib test
License: GPL
%description
%{summary}
%install
mkdir -p %{buildroot}/opt/colortest/
echo "%{_arch}" > %{buildroot}/opt/colortest/%{_arch}
%files
/opt/colortest/

$ rpmbuild -bb --target i686 colortest.spec
$ rpmbuild -bb --target x86_64 colortest.spec
```

Multilib has to be active (3):
```
$ rpm --eval '%_transaction_color'
3
```

And reproducer

```
# rpm -U colortest-1.0-1.i686.rpm
# rpm -q colortest
colortest-1.0-1.i686

# rpm -U colortest-1.0-1.x86_64.rpm
# rpm -q colortest
colortest-1.0-1.x86_64          ← i686 silently erased while it shouldn't be
```


So far findings are described here:

https://ixion.pld-linux.org/~arekm/rpm-multilib-bug/bug.html

and two claude derived patches:

https://ixion.pld-linux.org/~arekm/rpm-multilib-bug/rpm-patches/

I'm not posting these here because I don't have enough understanding of rpm internals to 100% confirm these findings and people just don't like AI bugreports without deep understanding of the code by human. The reproducer works well though.
